nixos/modules/services/databases/pgbouncer.nix

```diff
@@ -1,543 +1,138 @@
-      (mapAttrsToList (peerid : settings : "${peerid} = ${settings}") cfg.peers)}
-
-      [pgbouncer]
-      # general
-      ${optionalString (cfg.ignoreStartupParameters != null) "ignore_startup_parameters = ${cfg.ignoreStartupParameters}"}
-      listen_port = ${toString cfg.listenPort}
-      ${optionalString (cfg.listenAddress != null) "listen_addr = ${cfg.listenAddress}"}
-      pool_mode = ${cfg.poolMode}
-      max_client_conn = ${toString cfg.maxClientConn}
-      default_pool_size = ${toString cfg.defaultPoolSize}
-      max_user_connections = ${toString cfg.maxUserConnections}
-      max_db_connections = ${toString cfg.maxDbConnections}
-
-      #auth
-      auth_type = ${cfg.authType}
-      ${optionalString (cfg.authHbaFile != null) "auth_hba_file = ${cfg.authHbaFile}"}
-      ${optionalString (cfg.authFile != null) "auth_file = ${cfg.authFile}"}
-      ${optionalString (cfg.authUser != null) "auth_user = ${cfg.authUser}"}
-      ${optionalString (cfg.authQuery != null) "auth_query = ${cfg.authQuery}"}
-      ${optionalString (cfg.authDbname != null) "auth_dbname = ${cfg.authDbname}"}
-
-      # TLS
-      ${optionalString (cfg.tls.client != null) ''
-      client_tls_sslmode = ${cfg.tls.client.sslmode}
-      client_tls_key_file = ${cfg.tls.client.keyFile}
-      client_tls_cert_file = ${cfg.tls.client.certFile}
-      client_tls_ca_file = ${cfg.tls.client.caFile}
-      ''}
-      ${optionalString (cfg.tls.server != null) ''
-      server_tls_sslmode = ${cfg.tls.server.sslmode}
-      server_tls_key_file = ${cfg.tls.server.keyFile}
-      server_tls_cert_file = ${cfg.tls.server.certFile}
-      server_tls_ca_file = ${cfg.tls.server.caFile}
-      ''}
-
-      # log
-      ${optionalString (cfg.logFile != null) "logfile = ${cfg.homeDir}/${cfg.logFile}"}
-      ${optionalString (cfg.syslog != null) ''
-      syslog = ${if cfg.syslog.enable then "1" else "0"}
-      syslog_ident = ${cfg.syslog.syslogIdent}
-      syslog_facility = ${cfg.syslog.syslogFacility}
-      ''}
-      ${optionalString (cfg.verbose != null) "verbose = ${toString cfg.verbose}"}
-
-      # console access
-      ${optionalString (cfg.adminUsers != null) "admin_users = ${cfg.adminUsers}"}
-      ${optionalString (cfg.statsUsers != null) "stats_users = ${cfg.statsUsers}"}
-
-      # extra
-      ${cfg.extraConfig}
-    '';
-  };
-
-in {
+  settingsFormat = pkgs.formats.ini { };
+  configFile = settingsFormat.generate "pgbouncer.ini" cfg.settings;
+  configPath = "pgbouncer/pgbouncer.ini";
+in
+{
+  imports = [
+    (lib.mkRenamedOptionModule
+      [ "services" "pgbouncer" "logFile" ]
+      [ "services" "pgbouncer" "settings" "pgbouncer" "log_file" ])
+    (lib.mkRenamedOptionModule
+      [ "services" "pgbouncer" "listenAddress" ]
+      [ "services" "pgbouncer" "settings" "pgbouncer" "listen_addr" ])
+    (lib.mkRenamedOptionModule
+      [ "services" "pgbouncer" "listenPort" ]
+      [ "services" "pgbouncer" "settings" "pgbouncer" "listen_port" ])
```

Comment:
This gives me

```
error: evaluation aborted with the following error message: 'Renaming error: option `services.pgbouncer.settings.pgbouncer.listen_port' does not exist.'
```

btw.
Keeping options around is also useful for a painless upgrade path.